worlds/adventure/__init__.py

```diff
@@ -0,0 +1,355 @@
+            self.multiworld.itempool.append(self.create_item("Right Difficulty Switch"))
+            self.created_items += 1
+
+        extra_filler_count = num_locations - self.created_items
+        self.dragon_speed_reducer_info = {}
+        # make sure yorgle doesn't take 2 if there's not enough for the others to get at least one
+        if extra_filler_count <= 4:
+            extra_filler_count = 1
+        self.create_dragon_slow_items(self.yorgle_min_speed, self.yorgle_speed, "Slow Yorgle", extra_filler_count)
+        extra_filler_count = num_locations - self.created_items
+
+        if extra_filler_count <= 3:
+            extra_filler_count = 1
+        self.create_dragon_slow_items(self.grundle_min_speed, self.grundle_speed, "Slow Grundle", extra_filler_count)
+        extra_filler_count = num_locations - self.created_items
+
+        self.create_dragon_slow_items(self.rhindle_min_speed, self.rhindle_speed, "Slow Rhindle", extra_filler_count)
+        extra_filler_count = num_locations - self.created_items
+
+        # traps would probably go here, if enabled
+        freeincarnate_max = self.multiworld.freeincarnate_max[self.player].value
+        actual_freeincarnates = min(extra_filler_count, freeincarnate_max)
+        self.multiworld.itempool += [self.create_item("Freeincarnate") for _ in range(actual_freeincarnates)]
+        self.created_items += actual_freeincarnates
+
+        extra_filler_count = num_locations - self.created_items
+        self.multiworld.itempool += [self.create_item("nothing") for _ in range(extra_filler_count)]
+
+    def create_dragon_slow_items(self, min_speed: int, speed: int, item_name: str, maximum_items: int):
+        if min_speed < speed:
+            delta = speed - min_speed
+            if delta > 2 and maximum_items >= 2:
+                self.multiworld.itempool.append(self.create_item(item_name))
+                self.multiworld.itempool.append(self.create_item(item_name))
+                speed_with_one = speed - math.floor(delta / 2)
+                self.dragon_speed_reducer_info[item_table[item_name].id] = [speed_with_one, min_speed]
+                self.created_items += 2
+            elif maximum_items >= 1:
+                self.multiworld.itempool.append(self.create_item(item_name))
+                self.dragon_speed_reducer_info[item_table[item_name].id] = [min_speed]
+                self.created_items += 1
+
+    def create_regions(self) -> None:
+        create_regions(self.multiworld, self.player, self.dragon_rooms)
+
+    def set_rules(self) -> None:
+        set_rules(self)
+
+    def generate_basic(self) -> None:
+        self.multiworld.get_location("Chalice Home", self.player).place_locked_item(
+            self.create_event("Victory", ItemClassification.progression))
+        self.multiworld.completion_condition[self.player] = lambda state: state.has("Victory", self.player)
+
+    def pre_fill(self):
```

Comment:
If I'm understanding right, you're creating additional locations and items, then later remove them again? Why not just select a random set of locations and only create those?

Reply:
I thought I'd have problems with those unused locations either not showing in the tracker or being unable to 'check' them for the tracker.  And also I believe I need to make sure any locations that are plando'd exist, but I guess there's probably some way for me to get a list of those.  I'm placing items in those locations and removing those items from the pool, but I'll see what it looks like if I leave out those locations.  Or if that does have a problem, not creating the 'nothing' items in create_items and adding them to the pool and taking them back out later, just creating them as needed.  

Is removing items from the pool the slow part?  I don't want to slow down large multiworlds.

The largest one I tested was 2600 worlds and iirc took about 5 minutes, but that was almost all Adventure and the total location count was only in the tens of thousands, not like a real big async.  I'll note down the generation times as I try out these changes to make sure I don't make something awful.

That nothing item, if I keep it, can also be set up more consistently with the other extra filler items now, and I should go ahead and do that.

Reply:
Basically, as expected, they don't show up in the tracker if I don't create them.  Not generating those locations with random filler limits cuts total generation time to 70% of the original handling, and using all locations but skipping the item pool for any nothings reduces it to 90%.  That's with random amounts of filler; it is probably more typical for all locations to have a real item anyway, as that's the default setting.  There's just an option for limiting the number of freeincarnates that normally take up any extra slots, and if it runs out of those it uses 'nothing'.

Reply:
This is enough of a change I want to spend more time testing than I have right now, so I'll push this another day.  I'm having trouble staying awake now.

Reply:
It's rare for games to have this problem in the first place, generally players will prefer any trash item whatsoever over an actual nothing item. But in the cases where it can exist, mostly this is Super Metroid, they are actual Nothing items that as far as AP is concerned is just a regular item.

Reply:
As far as AP is concerned, they are Nothing items (ID 118000000) placed locally.

In Adventure, the built-in randomizer (level 3) has a lot more locations than vanilla items, so there are many empty rooms.  I want that to still be an option, without revealing what rooms are populated in the tracker, but still marking those as checked when they are seen.  In order to make it tracker-friendly, it adds 'Nothing' AP items to some locations which get checked when seen by the client.

The default yaml settings do generate a non-Nothing item for every location, but that changes the feel of exploring some, so I didn't want to enforce that as the way things have to be.